File: DOTNB_C/DOTNB.cpp
#include <iostream>
#include <cmath>
using namespace std;


// Note convergence restrictions: abs(x) < 1 and c not a negative integer or zero
double hypergeometric( double a, double b, double c, double x )
{
  const double TOLERANCE = 1.0e-10;
  double term = a * b * x / c;
  double value = 1.0 + term;
  int n = 1;
  
  while ( abs( term ) > TOLERANCE )
  {
    a++, b++, c++, n++;
    term *= a * b * x / c / n;
    value += term;
  }
  return value;
}

//Calculate mean of DOTNB distribution
double DOTNB_mean(int r1, double p1, int r2, double p2)
{
  double q1 = 1 - p1;
  double q2 = 1 - p2;
  double dmean = ((r1*q1)/p1) - (r2*q2/p2);
  return dmean;
}

//Calculate variance of DOTNB distribution
double DOTNB_var(int r1, double p1, int r2, double p2)
{
  double q1 = 1 - p1;
  double q2 = 1 - p2;
  double dvar = (r1*q1)/(pow(p1, 2)) + (r2*q2)/(pow(p2, 2));
  return dvar;
}

// calculate the p-value of two different NB distributions (r1,p1) and (r2,p2)
// dn is the observed number of difference
double DOTNB_pdf(int r1, double p1, int r2, double p2, int dn )
{
  double q1 = 1-p1, q2 = 1-p2;
  double prob;
  if(dn>0){
    prob = pow(p1,r1)*pow(p2,r2)*((tgamma(r1+dn)/tgamma(r1))/tgamma(1+dn))*pow(q1,dn)*hypergeometric(r1+dn,r2,dn+1,q1*q2);
  }else{
    dn=-dn;
    prob = pow(p1,r1)*pow(p2,r2)*((tgamma(r2+dn)/tgamma(r2))/tgamma(1+dn))*pow(q2,dn)*hypergeometric(r2+dn,r1,dn+1,q1*q2);
  }
  return prob;
}

//Calculate DOTNB cdf
double DOTNB_cdf(int r1, double p1, int r2, double p2, int dn, int maxiter){
  double cvalue = 0;
  double pdfv = DOTNB_pdf(r1, p1, r2, p2, dn);
  int iter = 0;
  dn = round(dn);
  float TOLERANCE = 1.0e-20;
   
  if(pdfv < TOLERANCE){
     
    if (pdfv > DOTNB_mean(r1, p1, r2, p2)){
      cvalue = 1;
    }
    else{
      cvalue = 0;
    }
     
  }else{
     
    while(pdfv > TOLERANCE){
      cvalue=cvalue+pdfv;
      dn = dn - 1;
      pdfv=DOTNB_pdf(r1,p1,r2,p2,dn);
      iter = iter + 1;
      if(iter > maxiter){
        break;
      }
    }
     
  }
  return cvalue;
}

//======================================================================

// Here is an example to calculate the DOTNB_pdf(r1=2,p1=0.58,r2=3,p2=0.3,k=10)
int main()
{
   int r1=2, r2=3,k=10;
   double p1=0.58,p2=0.3;
   cout<<"The probability of the number "<<k<< " of difference occurring is "<< DOTNB_pdf(r1,p1,r2,p2,k)<<"\n";
}

File: DOTNBR/R/DOTNB.R
#'@title DOTNB_pdf
#'@description
#'Calculates the p-value for the probability of significant difference between two negative binomial distributions using the DOTNB probability density function.
#'
#'@param r1 The r parameter for the first distribution
#'@param p1 The p parameter for the first distribution
#'@param r2 The r parameter for the second distribution
#'@param p2 The p parameter for the second distribution
#'@param tt the observed difference between the two distributions
#'@return the PDF output
#'@export

DOTNB_pdf<-function(r1,p1,r2,p2,tt){
  q1 = 1-p1
  q2 = 1-p2
  if(tt>0){
    dd = (p1^r1)*(p2^r2)*((gamma(r1+tt)/gamma(r1))/gamma(1+tt))*(q1^tt)*genhypergeo(c(r1+tt,r2),tt+1,q1*q2);
  }else{
    ntt=-tt
    dd = (p1^r1)*(p2^r2)*((gamma(r2+ntt)/gamma(r2))/gamma(1+ntt))*(q2^ntt)*genhypergeo(c(r2+ntt,r1),ntt+1,q1*q2);
  }
  return(dd)
}

#'@title DOTNB_mean
#'@description
#'Calculates the mean for the DOTNB distribution based on the parameters of two negative binomial distributions
#'
#'@param r1 The r parameter for the first distribution
#'@param p1 The p parameter for the first distribution
#'@param r2 The r parameter for the second distribution
#'@param p2 The p parameter for the second distribution
#'@return The mean for the representative DOTNB distribution
#'@export
DOTNB_mean<-function(r1,p1,r2,p2){
  q1 = 1-p1
  q2 = 1-p2
  dmean = ((r1*q1)/p1) - (r2*q2/p2)
  return(dmean)
}

#'@title DOTNB_mean
#'@description
#'Calculates the variance for the DOTNB distribution based on the parameters of two negative binomial distributions
#'
#'@param r1 The r parameter for the first distribution
#'@param p1 The p parameter for the first distribution
#'@param r2 The r parameter for the second distribution
#'@param p2 The p parameter for the second distribution
#'@return The variance for the representative DOTNB distribution
#'@export
DOTNB_var<-function(r1,p1,r2,p2){
  q1 = 1-p1
  q2 = 1-p2
  dvar = (r1*q1)/(p1^2) + (r2*q2)/(p2^2)
  return(dvar)
}

#'@title DOTNB_pdf
#'@description
#'Calculates the p-value for the probability of significant difference between two negative binomial distributions using the DOTNB cumulative density function.
#'
#'@param r1 The r parameter for the first distribution
#'@param p1 The p parameter for the first distribution
#'@param r2 The r parameter for the second distribution
#'@param p2 The p parameter for the second distribution
#'@param tt the observed difference between the two distributions
#'@return the CDF output
#'@export
DOTNB_cdf<-function(r1,p1,r2,p2,tt){
  cvalue=0
  pdfv=DOTNB_pdf(r1,p1,r2,p2,tt)
  maxiter = 2000
  if(is.infinite(pdfv)){
    if (tt > DOTNB_mean(r1,p1,r2,p2)){
      cvalue=1
    }
  }else{
    iter = 1
    while(is.finite(pdfv)){
      cvalue=cvalue+pdfv
      tt=tt-1
      pdfv=DOTNB_pdf(r1,p1,r2,p2,tt)
      iter = iter +1
      if(iter > maxiter){
        break
      }
    }
  }
  return(cvalue)
}

